### DISCUSSION TITLE
thin mode not pure python

### DISCUSSION THREAD

**Discussion by emdete** [oracle/python-oracledb#87] [General]
i would expect thin mode to be implemented purely in python but found a shared library thin_impl. this makes it impossible to run the code on different platforms with python support and is not the idea of "thin" according to my understanding.

**Comment by anthony-tuininga** [comment]
I can understand why you would like to have a pure Python implementation but this is not practical at this point -- unless of course performance is not a concern! There are different definitions of "thin". The one we are using is the one that indicates that nothing is required besides the driver itself to connect to the Oracle Database. No additional client libraries are required. The `thin_impl` shared library is written in Cython and so is converted from Python to C code for performance reasons. You should be able to use the driver on any platform that supports Python (and the cryptography package) -- you'll simply have to build it for platforms for which we don't supply binaries.

**Comment by cjbj** [comment]
One reference point is that Oracle's 'thin' JDBC driver is not released as Java source.

**Comment by emdete** [comment]
how is the closed source topic of the java driver related, that one is pure java and named thin as well. and sure  - you can always redefine the meaning of words.

but i read that performance is the cause for the native code?

**Comment by cjbj** [comment]
What name would you like to give the mode?

**Comment by cjbj** [comment]
@emdete I moved this to a 'Discussion'.  I'd be happy to hear how we can improve our messaging about python-oracledb.

**Reply by emdete** [reply]
thin and thick type of database adapters showed up with java as far as i remember. if i'm not wrong thick mode meant that native binaries where needed while thin meant a pur java implementation (making deployment easier). back then oracle provided a single jar which could do both, thick mode only for one platform.

because the python implementation contains a share library, or better native code i would think that thick would be the correct term if one want to go with that terminology and deployment constraints (i cannot copy my installation from linux to solaris for example).

on the other hand you are right, the implementation is "complete", everything is implemented in the python world, i think there is no way but to explain this special design.
